keras/03-the-model-functional-and-sequential-class.md

```diff
@@ -0,0 +1,250 @@
+`node2` and `node5` are in `layer_b._inbound_nodes`.
+
+`node1 -> layer_a -> node2 -> layer_b -> node3`
+
+`node4 -> layer_a -> node5 -> layer_b -> node6`
+
+When building a functional model, we can call a `Layer` with multiple
+`KerasTensor`s. For example, the
+[`Add`](https://keras.io/api/layers/merging_layers/add/) layer add multiple
+tensors together. Therefore, a call of a layer corresponds to multiple
+`KerasTensor`s. A `Node` also corresponds to a call of a `Layer`. Therefore,
+a `Node` may correspond to multiple `KerasTensor`s. A `Node` record these
+`KerasTensor`s in `Node._keras_inputs`.
+
+`KerasHistory` is for the `KerasTensor` to find the input `Layer` and the
+corresponding `Node`. It is stored in the attribute of
+`KerasTensor._keras_history`. `KerasHistory.layer` records the `Layer`
+producing it. `KerasHistory.node_index` records the index of the corresponding
+`Node` in the `KerasHistory.layer._inbound_nodes`.
+
+For example, if `node2` has a corresponding `KerasTensor`, named
+`keras_tensor_2`, `keras_tensor_2._keras_history.node_index` records the index
+of `node1` in `layer_a._inbound_nodes`.
+
+With all these recording mechanisms, we can have the following pseudo-code.
+
+([Source](https://github.com/keras-team/keras/blob/v2.6.0/keras/engine/base_layer.py#L2591))
+
+```py
+class Layer(tf.Module):
+  def __call__(inputs):
+    ...
+    outputs = self.call(inputs)
+    node = Node(self, inputs, outputs)
+    ...
+```
+
+([Source](https://github.com/keras-team/keras/blob/v2.6.0/keras/engine/node.py#L96-L100))
+
+```py
+class Node:
+  def __init__(self, layer, inputs, outputs):
+    ...
+    self._keras_inputs = inputs
+    layer._inbound_nodes.append(self)
+    node_index = len(self.layer._inbound_nodes) - 1
+    for keras_tensor in outputs:
+        keras_tensor._keras_history = KerasHistory(layer, node_index)
+    ...
+```
+
+Now, we have the answers to the two questions above. A `KerasTensor` find the
+layer producing it with `KerasTensor._keras_history.layer`. A `Layer` find the
+input `KerasTensor` with `Layer._inbound_nodes[0]._keras_inputs`.
+
+Finally, we can fetch the entire computational graph with the algorithm
+described above. The pseudo-code is shown as follows.
+
+([Source](https://github.com/keras-team/keras/blob/v2.6.0/keras/engine/functional.py#L1031))
+
+```py
+def fetch_nodes(output_keras_tensor):
+  if output_keras_tensor in model.inputs:
+    return []
+  layer = output_keras_tensor._keras_history.layer
+  node_index = output_keras_tensor._keras_history.node_index
+  node = layer._inbound_nodes[node_index]
+  node_list = [node]
+  for input_keras_tensor in node._keras_inputs:
+    node_list += fetch_nodes(input_keras_tensor)
+  return node_list
+```
+
+The actual code would not only fetch the nodes, but also the layers, and sort
+them in topological order. In `Functional.call`, it calls the layer in
+topological order to produce the outputs.
+
+### The `Sequential` class
```

Comment:
It would be useful to talk about the two types of Sequential models: deferred-build (variables and underlying Functional model aren't created until the model is called) and Functional-like (input shapes are known from the start and the underlying Functional model exists from the start).

Also link to the guide about the Sequential model.